klippy/extras/load_cell_probe.py

```diff
@@ -0,0 +1,611 @@
+    cmd_LCP_CALIB_NOISE_help = "Determine typical noise level of load "        \
+            + "cell measurements."
+    cmd_LCP_CALIB_WEIGHT_help = "Calibrate to physical weight unit."
+    cmd_LCP_CALIB_STIFFNESS_help = "Calibrate system stiffness."
+    cmd_LCP_INFO_help = "Print load cell probe parameters to console."
+
+
+    def get_lift_speed(self, gcmd=None):
+        if gcmd is not None:
+            return gcmd.get_float("LIFT_SPEED", SPEED, above=0.)
+        return SPEED
+
+
+    def get_offsets(self):
+        return 0, 0, 0
+
+
+    def multi_probe_begin(self):
+        pass
+
+
+    def multi_probe_end(self):
+        pass
+
+
+    def get_status(self, eventtime):
+        return {
+          'last_force': self._last_force,
+          'last_z_result': self._last_z_result,
+        }
+
+
+    def run_probe(self, gcmd):
+
+        # wait until toolhead is in position
+        self.tool.wait_moves()
+
+        repeat_count=0
+        while True:
+          # fast, coarse approach
+          force = self._fast_approach(gcmd)
+
+          # find start position for fit (no contact but as close to surface)
+          self._find_fit_start(gcmd, force)
+
+          # perform raster scan and fit
+          result = self._perform_fit(gcmd)
+          if result is not None:
+            break
+
+          # check abort condition
+          repeat_count += 1
+          if repeat_count > MAX_RETRY:
+            raise gcmd.error("Maximum retries reached, giving up.")
+          gcmd.respond_info("Retrying...")
+          self._move_z_relative(5*self._step_size, False)
+
+        pos = self.tool.get_position()
+        gcmd.respond_info("FINISHED toolhead Z = %f" % result)
+        pos[2] = result
+        self._last_z_result = result
+        return pos[0], pos[1], pos[2]
+
+
+    def _handle_ready(self):
+        # obtain toolhead object
+        self.tool = self._printer.lookup_object('toolhead')
+
+
+    def _adc_callback(self, time, value):
+        # convert to physical unit
+        self._last_uncompensated_force = value * self._force_calibration
+        # First value after start: use as zero offset
+        if self._force_offset == None :
+          self._force_offset = self._last_uncompensated_force
+        self._last_time = time
+        # Store zero offset compensated value for display
+        self._last_force = self._last_uncompensated_force - self._force_offset
+
+
+    def _adc_wait_conversion_ready(self, gcmd):
+        last_time = self._last_time
+        clocksync = self._mcu_adc.get_mcu()._clocksync
+        clock = clocksync.print_time_to_clock(last_time)
+        last_sys_time = clocksync.estimate_clock_systime(clock)
+        for n in range(1,10) :
+          # wait shortly after the timer has called _sample_timer
+          self._reactor.pause(last_sys_time + n*self._report_time + 0.0001)
+          if self._last_time != last_time:
+            return
+        # callback not called after 10 report time intervals -> error
```

Comment:
I'm not sure what this code is attempting to do, but it does not look correct.

At a minimum, the code must not "peek into" the member variables of other classes (`get_mcu()._clocksync` is not valid).  This is described briefly in docs/Code_Overview.md .